executor/pkg/executor/cl_executor.go

```diff
@@ -297,3 +281,75 @@ func (cle *ChainlinkExecutor) Validate() error {
 	}
 	return nil
 }
+
+// GetMessageStatus checks if a message should be executed and/or retried.
+// Returns (shouldRetry bool, shouldExecute bool, error) to indicate whether the message should be retried (added back to heap) and executed.
+func (cle *ChainlinkExecutor) GetMessageStatus(ctx context.Context, message protocol.Message, currentTime int64) (executor.MessageStatusResults, error) {
+	messageID, err := message.MessageID()
+	if err != nil {
+		return executor.MessageStatusResults{}, fmt.Errorf("failed to get message ID: %w", err)
+	}
+	if cle.isCursed(message) {
+		cle.lggr.Infow("Lane is cursed, skipping execution for message", "messageID", messageID)
+		return executor.MessageStatusResults{ShouldRetry: true, ShouldExecute: false}, nil
+	}
+	return cle.GetExecutionState(ctx, message, messageID)
+}
+
+func (cle *ChainlinkExecutor) isCursed(message protocol.Message) bool {
+	// todo: implement
+	return false
+}
+
+// GetExecutionState checks the onchain execution state of a message and returns if it should be retried and executed.
+// It does not do any checks to determine if verifications are available or not.
+// MESSAGE_UNTOUCHED: Message should be executed and retried.
+// MESSAGE_IN_PROGRESS: Message reentrancy protection, should not be retried, should not be executed.
+// MESSAGE_SUCCESS: Message was executed successfully, don't retry and don't execute.
+// MESSAGE_FAILURE: Message failed to execute, don't retry and don't execute.
+func (cle *ChainlinkExecutor) GetExecutionState(ctx context.Context, message protocol.Message, id protocol.Bytes32) (ret executor.MessageStatusResults, err error) {
+	// Check if the message is already executed to not waste gas and time.
+	destinationChain := message.DestChainSelector
+
+	executionState, err := cle.destinationReaders[destinationChain].GetMessageExecutionState(
+		ctx,
+		message,
+	)
+	if err != nil {
+		// If we can't get execution state, don't execute, but put back in heap to retry later.
+		return executor.MessageStatusResults{ShouldRetry: true, ShouldExecute: false}, fmt.Errorf("failed to check GetMessageExecutionState: %w", err)
+	}
+	switch executionState {
+	case executor.MESSAGE_SUCCESS:
+		ret.ShouldRetry = false
+		ret.ShouldExecute = false
+		err = nil
+
+		// In progress is a status code for reentrancy protection
+	case executor.MESSAGE_IN_PROGRESS:
+		ret.ShouldRetry = false
+		ret.ShouldExecute = false
+		err = nil
+
+	// If message has failed to execute in the past, don't retry and don't execute.
+	// Any changes to VerifierResults should be manually executed.
+	case executor.MESSAGE_FAILURE:
+		ret.ShouldRetry = false
+		ret.ShouldExecute = false
+		err = nil
```

Comment:
More code here than needed. You can combine these cases to make it clear that htey all have the same handling
```
	case executor.MESSAGE_SUCCESS,
		executor.MESSAGE_IN_PROGRESS,
		executor.MESSAGE_FAILURE:
		ret.ShouldRetry = false
		ret.ShouldExecute = false
		err = nil
```

```
	case executor.MESSAGE_SUCCESS:
		fallthrough
	case executor.MESSAGE_IN_PROGRESS:
		fallthrough
	case executor.MESSAGE_FAILURE:
		ret.ShouldRetry = false
		ret.ShouldExecute = false
		err = nil
```

Or just put it in the default:
```
	// We only execute untouched messages
	case executor.MESSAGE_UNTOUCHED:
		ret.ShouldRetry = true
		ret.ShouldExecute = true
		err = nil
	default:
		ret.ShouldRetry = false
		ret.ShouldExecute = false
		err = nil
```